README.md

### photoSpec: A photographic spectrophotometer

At this time, `photoSpec is a WORK IN PROGRESS`.  It has matured quite a bit but is still undergoing tests which may significantly change the code without warning.

photoSpec allows one to apply Beer's Law to a photograph containing color swatches (paint chips) and a sample to be analyzed.  The motivation is to provide a method for 'spectrophotometry' of objects that cannot be chemically analyzed because of their value or non-extractability.  Functions in the package cover the entire workflow from generating calibration colors to obtaining a quantitative answer analogous to concentration in Beer's Law.

#### To install from Github using R:

````r
install.packages("devtools")
library("devtools")
install_github(repo = "photoSpec", username = "bryanhanson", ref = "master")
library("photoSpec")
````
If you use `ref = "devel"` you can get the development branch if it is available.  Development branches have new, possibly incompletely tested features.  They may also not be ready to pass checks and thus may not install automatically using the method above.  Check the news file to see what's up.

Questions?  hanson@depauw.edu

NEWS

NEWS file for package photoSpec
photoSpec: A photographic spectrophotometer
Bryan A. Hanson DePauw University, Greencastle Indiana USA

##### Version 0.5-4
Date: 2014-03-27
>> Renamed several internalFunctions to fuller, more descriptive names: ccp -> computeColorPurity; dAB -> distAB; makecC -> makeCalCols; xy2cC -> CIExy2calCols.
>> Forgot to export useArbCols, now added.  Also needed documentation file.
>> Added documentation on calCols structure.
>> Implemented sort option in makeCalCols.
>> Added function sortColors to internalFunctions.

##### Version 0.5-3
Date: 2014-03-27
>> Brought selectCalCols back from version 3.0.  It is used with selectCIExy to convert the CIE xy coordinates to a limited number of Munsell colors for use in the cards.  Thus one can now use two wavelengths with the CIE diagram to select your calibration colors, which is much closer to the spirit of how one would use spectroscopic information to derive the appropriate colors.
>> selectCIExy renamed to genColsUsingLambdas.
>> selectCalCols renamed to xy2cC and moved to internalFunctions.
>> Related to the above, genColsUsingLambdas now gains a res (resolution) argument which determines how the upper number of Munsell colors are selected from w/i the 'wedge'.
>> Added prepGrid to internalFunctions to centralize this task.

>> All functions returning a calCols object (gen5cols, genManyCols, genColsUsingLambdas, and  useArbCols) now return a data frame instead of a list.
>> Options to display results made consistent between gen5cols, genManyCols, genColsUsingLambdas, and useArbCols.
>> Added function makecC to internalFunctions which constructs the calCols in a consistent way, and can start from any color format.

>> In cleaning up code, function grid2rgb was added to internalFunctions.  It takes care of a job that was common to prepCIEgradient and the old selectCalCols.
>> prepCIEgradient moved to internalFunctions.

##### Version 0.5-2
Date: 2014-02-08
>> Fixed Munsell color labeling in plotCIEchrom.
>> Added a check for calCols in showRGBcalibration.  Stops if not provided.
>> Made calCols optional in showCIE, so that one can focus on the sampCol plotting if desired.  Actually, this was possible already but made it clear in the documentation.
>> Added a plot which defines color purity to the documentation of calcColorPurity.

##### Version 0.5-1
Date: 2014-02-02
>> Lowered the resolution of the CIExyz data set from every 0.1 nm to every 1.0 nm.  This greatly improved the speed of the package.  A lot of code changes were necessary.
>> The method used to plot the lambda lines in calcColorPurity was changed significantly, as the lower resolution data required interpolating some points in order for the lines to pass through the points they were supposed to pass through.

##### Version 0.4-18
Date: 2013-12-23
>> Further tweaks of the documentation.
>> Lots of troubleshooting in findPolygonIntersection, called by ccp.  Trying to find the issue in which the number of gamut intersections doesn't match the number of sample numbers given.  It is now certain it arises from a zero-length segment, but I'm not sure why this occurs.  Improved error messages.  Work-around is to move lambdas 1 unit either way.

##### Version 0.4-17
Date: 2013-12-20
>> findCIEindex2, the vectorized version, was renamed to findCIEindex.
>> Removed a lot of temporary commenting/alternate code sections.
>> Organized the functions in internalFunctions.
>> Improved documentation.

##### Version 0.4-16
Date: 2013-12-20
>> calcSampleValue return value 'value' is now called 'MunVal'.
>> Added function plotMiniCard.
>> Added function gen5cols for use with plotMiniCard.
>> Function plotSampleCard renamed to plotLargeCard for consistency with plotMiniCard.
>> Function genCalCols renamed to genManyCols so that its relation to gen5cols is clear.
>> Function showCIEcalibration renamed to showCIE, since there is no true calibration.  Function also modified so that any combination of calCols and sampCols can be displayed.
>> The conversion of hex codes to rgb and then to CIE x,y coordinates was moved to internalFunctions since it appeared in several functions.
>> Removed calls to findCIEindex and replaced them with findCIEindex2.


##### Version 0.4-15
Date: 2013-11-16
>> Previous attempt to fix ccp did not work!  It now correctly computes color purity, including negative values.
>> Modified calcColorPurity and ccp for increased speed and ease of code maintenance.
>> Created findCIEindex2, a vectorized version of findCIEindex.  Have not yet systematically changed all code to use it yet, however.
>> Vectorized lineIntersection and updated code which uses it.
>> Vectorized dAB and updated code which uses it.

##### Version 0.4-14
Date: 2013-11-12
>> Modified ccp to return negative color purity values where appropriate.

##### Version 0.4-13
Date: 2013-10-30
>> Additional improvements to the documentation, especially the examples.
>> Moved getRGorB to internalFunctions.R
>> Use of getRGorB moved to runAllCal from calcColorPurity.
>> Argument sampCol, in several functions, was renamed to sampCols to be consistent with calCols.

##### Version 0.4-12
Date: 2013-10-24
>> Semantic versioning is now in effect (it basically was already, this makes it official).  See semver.org
>> Small fixes to documentation.
>> Added a helper function, getRGorB to calcColorPurity.  This simply extracts the R, G or B channel from the sampCol vector and adds it to the return value.

##### Version 0.4-11
Date: 2013-08-22
>> Fixed a problem in calcColorPurity in which a segment from a wavelength through D65 which should hit the line of purples did not.
>> Another major trek through the documentation, including better examples.

##### Version 0.4-10
Date: 2013-08-12
>> Fixed a problem in calcColorPurity in which a segment from a wavelength through D65 which should hit the line of purples did not.
>> Changed colors and symbols on calcColorPurity when plotLambdas = TRUE.

R/plotCIEchrom.R

plotCIEchrom <- function(gradient = NULL, colSpace = "sRGB", ex = 1.0,
	opts = c("D65", "specLocus", "purples"), title = NULL, ...) {

	# Function to draw the CIE chromaticity diagram
	# with various decorations
	
	# Bryan Hanson, DePauw University, March 2013 hanson@depauw.edu
	# Part of the photoSpec package
	
##### Get and prepare the data
	# These are the coordinates of the spectral locus, which is a curve
	# describing the pure colors of the spectrum/rainbow

	data("CIExyz", envir = environment())
	# Note z = 1- x - y
	# Cutoff the data at 650 nm; beyond that the curve strangely
	# turns back on itself
	Lxyz <- subset(CIExyz, CIExyz$wavelength <= 650)
		
##### Prepare the raster with the color gradient
	
	if (!is.null(gradient)) {
		if (!(colSpace == "sRGB") || (colSpace == "Apple RGB")) stop("colSpace must be sRGB or Apple RGB")
		if (is.character(gradient)) { # captures, crudely, gradient = "sl"
			finras <- as.raster(prepCIEgradient(vertices = Lxyz,
				colSpace = colSpace, ex = ex, ...))
			}
			
		if (is.data.frame(gradient)) {
			finras <- as.raster(prepCIEgradient(vertices = gradient,
				colSpace = colSpace, ex = ex, ...))
			}
		}
	
##### Create the plot using grid graphics

	off <- 0.1 # needed due to vp origin at -0.1
	Lxyz$x <- Lxyz$x + off
	Lxyz$y <- Lxyz$y + off

	# First plot titles & labels in the vp of the entire device

	if (is.null(title)) title <- "1931 CIE Chromaticity Diagram"
	grid.newpage()
	grid.text(title, x = 0.5, y = 0.9,
		gp = gpar(fontface = "bold", cex = 1.2))
	grid.text(expression(italic(x)), x = 0.5, y = 0.05)
	grid.text(expression(italic(y)), x = 0.05, y = 0.5, rot = 90)

	# Now the data in it's own viewport; raster goes underneath (first)

	pushViewport(viewport(width = 0.7, height = 0.7,
		xscale = c(-0.1, 0.9), yscale = c(-0.1, 0.9)))

	if (!is.null(gradient)) {
		grid.raster(finras, x = 0.5, y = 0.5, interpolate = FALSE, default.units = "npc")
		msg1 <- "Warning: the color gradient appearance\nwill vary with the device, surface\n& incident light used to view it\nand is not likely correct anywhere"
		grid.text(msg1, x = 0.98, y = 0.9, gp = gpar(fontface = "italic", cex = 0.9),
			just = "right")
		msg2 <- paste("ex =", ex, "    color space =", colSpace, sep = " ")
		grid.text(msg2, x = 0.98, y = 0.02, gp = gpar(cex = 0.5),
			just = "right")
		}

	grid.polygon(Lxyz$x, Lxyz$y)
	grid.rect()
	tickpos <- seq(0.0, 0.8, by = 0.1)
	grid.xaxis(at = tickpos)
	grid.yaxis(at = tickpos)

	# Labels for the spectral locus
	
	specl <- c(11, 86, 111, 136, 161, 186, 211, 236, 261)
	labs <- c("400 nm  ", "475 nm  ", "500 nm  ", "  525 nm", "  550 nm",
		"  575 nm", "  600 nm", "  625 nm", "  650 nm")
	
	grid.points(x = Lxyz$x[specl], y = Lxyz$y[specl], gp = gpar(col = "black"),
		size = unit(0.5, "char"), default.units = "npc")
	grid.text(label = labs, Lxyz$x[specl], Lxyz$y[specl],
		hjust = c(1, 1, 1, 0, 0, 0, 0, 0, 0),
		vjust = c(1, 0, 0, 0, 0, 0, 0, 0, 1),
		gp = gpar(cex = 0.75))
	
	# Optional labeling
	
	# Need to add lty and a legend to the various gamut options
	
	# White point labels
	
	if ("D65" %in% opts) {
		wh <- getWhiteValues("D65")
		grid.points(wh$x, wh$y, gp = gpar(col = "black"),
			size = unit(0.5, "char"), default.units = "native")
		grid.text(wh$x, wh$y, label = "  D65", just = "left",
			gp = gpar(cex = 0.75), default.units = "native")
		}

	if ("D50" %in% opts) {
		wh <- getWhiteValues("D50")
		grid.points(wh$x, wh$y, gp = gpar(col = "black"),
			size = unit(0.5, "char"), default.units = "native")
		grid.text(wh$x, wh$y, label = "  D50", just = "left",
			gp = gpar(cex = 0.75), default.units = "native")
		}

	if ("C" %in% opts) {
		wh <- getWhiteValues("C")
		grid.points(wh$x, wh$y, gp = gpar(col = "black"),
			size = unit(0.5, "char"), default.units = "native")
		grid.text(wh$x, wh$y, label = "  C", just = "left",
			gp = gpar(cex = 0.75), default.units = "native")
		}

	if ("E" %in% opts) {
		wh <- getWhiteValues("E")
		grid.points(wh$x, wh$y, gp = gpar(col = "black"),
			size = unit(0.5, "char"), default.units = "native")
		grid.text(wh$x, wh$y, label = "  E", just = "left",
			gp = gpar(cex = 0.75), default.units = "native")
		}

	# gamut outlining
	
	if ("sRGB" %in% opts) {
		g <- getGamutValues("sRGB")
		grid.polygon(g$x, g$y, default.units = "native")
		}

	if ("SWOP" %in% opts) {
		g <- getGamutValues("SWOP")
		grid.polygon(g$x, g$y, default.units = "native")
		}

	if ("Apple" %in% opts) {
		g <- getGamutValues("Apple")
		grid.polygon(g$x, g$y, default.units = "native")
		}

	if ("NTSC" %in% opts) {
		g <- getGamutValues("NTSC")
		grid.polygon(g$x, g$y, default.units = "native")
		}

	if ("Adobe" %in% opts) {
		g <- getGamutValues("Adobe")
		grid.polygon(g$x, g$y, default.units = "native")
		}

	if ("CIE" %in% opts) {
		g <- getGamutValues("CIE")
		grid.polygon(g$x, g$y, default.units = "native")
		}

	# Misc decorations
	
	if ("Munsell" %in% opts) {
		mun <- c(81, 99, 112, 123, 166, 191, 211, 266)
		labs <- c("PB  ", "B  ", "BG  ", "G", "  GY",
			"  Y", "YR", "R")
		
		# grid.points(x = Lxyz$x[sl], y = Lxyz$y[sl], gp = gpar(col = "black"),
			# size = unit(0.5, "char"), default.units = "npc")
		grid.text(label = labs, Lxyz$x[mun], Lxyz$y[mun],
			hjust = c(1, 1, 1, 1, 0, 0, 0.5, 0),
			vjust = c(1, 1, 0, -1, 0, 0, -1.5, 2),
			gp = gpar(cex = 0.75))
		}

	if ("specLocus" %in% opts) {
		grid.text(0.75, 0.55, label = "spectral\nlocus", just = "left",
			gp = gpar(cex = 0.75), default.units = "native")
		grid.segments(0.73, 0.53, 0.61, 0.41, default.units = "native",
			arrow = arrow(ends = "last", length = unit(0.025, "npc"),
			angle = 15, type = "closed"))
		}


	if ("purples" %in% opts) {
		grid.text(0.65, 0.05, label = "line of\npurples", just = "left",
			gp = gpar(cex = 0.75), default.units = "native")
		grid.segments(0.63, 0.07, 0.5, 0.17, default.units = "native",
			arrow = arrow(ends = "last", length = unit(0.025, "npc"),
			angle = 15, type = "closed"))
		}

	if (is.null(gradient)) ans <- NULL
	if (!is.null(gradient)) ans <- finras
	invisible(ans)
	}